### PR TITLE
Update TinyMCE to the latest version

### PR DESCRIPTION
**Pull request by ellatrix**
## Description
This PR updates TinyMCE (4.6.5 to 4.7.1).
Changes: https://github.com/tinymce/tinymce-dist/compare/4.6.5...4.7.1

## How Has This Been Tested?
Make sure editable regions work, check things like paste, formatting, block splitting and merging...

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

**Comment by iseulde** [comment]
There's no good fix for the errors without addressing it in TinyMCE: https://github.com/tinymce/tinymce/issues/3947

**Comment by aduth** [comment]
FYI 4.7.0 was released today:

https://github.com/tinymce/tinymce-dist/commit/188f36662f3e31a3954db7aa5382533d6b64207c

**Comment by iseulde** [comment]
Yep :)

**Comment by iseulde** [comment]
Does anyone know why this is a dependency in `package.json`?

**Comment by youknowriad** [comment]
@iseulde maybe unit tests?

**Comment by iseulde** [comment]
Am I missing something? https://github.com/WordPress/gutenberg/search?utf8=✓&q=import+tinymce&type=

Is it the Editable component in a test? Why is it not the some TinyMCE file?

**Comment by iseulde** [comment]
So now the tests have undefined global errors. That's not supposed the run in node...

**Comment by aduth** [comment]
The tests run entirely in Node. Even if the tests don't explicitly import TinyMCE, the files under test may do so (either directly or indirectly). While in the browser the `import` statements are [converted to point to the window global](https://github.com/WordPress/gutenberg/blob/4bc2f885857dcf59765be00b8e3cde89c2e6a87e/webpack.config.js#L60), in the test environment they'd need the dependency to be available in `node_modules`. There are some issues with globals, notably in that the tests are set up to use JSDOM which is in many ways imperfect and doesn't implement all of the browser APIs. It appears this is now working well after adding to `test/setup-globals.js`?

**Comment by iseulde** [comment]
@aduth Yep, adjusted that file. Thanks. I was more wondering why we don't use the NPM module for the browser as well, instead of pulling it from a CDN.

To me this branch looks good to merge.

File: test/setup-globals.js [hunk 1]
```diff
@@ -19,6 +19,7 @@ global.window.tinyMCEPreInit = {
 };
 window.requestAnimationFrame = setTimeout;
 window.cancelAnimationFrame = clearTimeout;
+window.matchMedia = () => ( {} );
```
Comment (aduth) [inline test/setup-globals.js]: I wonder the extent to which we should try to mimic a proper return value here, i.e. [MediaQueryList](https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList). This implementation _could_ be problematic if a library using `matchMedia` attempted to treat the `matches` property on the return object as a boolean (i.e. `false === window.matchMedia( '...' ).matches`). Conversely, the value wouldn't be accurate, so we can't provide a useful value anyways. From what I can tell, this should be fine enough for how TinyMCE is using it:

https://github.com/tinymce/tinymce/blob/114e2d6878c58d6fe452db39353cc177f842d8bd/src/themes/mobile/src/main/js/touch/view/Orientation.js#L22

A more thorough stub might look like:

https://github.com/WickyNilliams/enquire.js/issues/82#issuecomment-26990494

File: blocks/editable/index.js [hunk 2]
```diff
@@ -119,7 +119,7 @@ export default class Editable extends Component {
 		editor.on( 'keyup', this.onKeyUp );
 		editor.on( 'selectionChange', this.onSelectionChange );
 		editor.on( 'BeforeExecCommand', this.maybePropagateUndo );
-		editor.on( 'BeforePastePreProcess', this.onBeforePastePreProcess );
+		editor.on( 'PastePreProcess', this.onPastePreProcess, true /* Add before core handlers */ );
```
Comment (aduth) [inline blocks/editable/index.js]: Why was this change necessary?

**Review by aduth — COMMENTED** [review]
One regression I'm noting: The "New Paragraph" placeholder text is showing as a darker color than that of master.

`master`|`update/tinymce`
---|---
master|branch

**Comment by aduth** [comment]
I updated the `package-lock.json` file with the new dependency, but also seeing some warnings and unrelated changes to integrity I wouldn't expect. Might need to...revisit ([maybe this known issue](https://github.com/npm/npm/issues/17749)).

**Comment by aduth** [comment]
>Thanks. I was more wondering why we don't use the NPM module for the browser as well, instead of pulling it from a CDN.

It probably doesn't make much sense when we're running a more bleeding-edge version than what's used elsewhere in Core, but the general point with aliasing to window globals is to play more nicely with shared dependencies of script loader, rather than potentially including multiple copies of the same dependency on the page by bundling it.

File: blocks/editable/index.js (hunk 2 above)
Comment (iseulde) [reply blocks/editable/index.js]: The `BeforePastePreProcess` event has been removed.

File: test/setup-globals.js (hunk 1 above)
Comment (iseulde) [reply test/setup-globals.js]: Sure. Should we adjust it here to something like that comment?

**Comment by iseulde** [comment]
Looks like there's an extra div inserted.

screenshot 2017-10-04 17 54 01

File: test/setup-globals.js (hunk 1 above)
Comment (aduth) [reply test/setup-globals.js]: I think it'd be an improvement, though in whatever solution we use, the result will be inaccurate at best. The implementation in the comment could at least avoid some strange behavior that could occur when trying to use properties of the return value (e.g. `addListener` as a noop at least prevents errors calling undefined as a function)

**Comment by youknowriad** [comment]
This will close #1607

**Comment by iseulde** [comment]
The placeholder issue is coming from https://github.com/tinymce/tinymce/commit/40ff87c0bf5bb610c560a0a14d9a8b67bacb47b8. Maybe we should add a class there after all, but don't really like TinyMCE inserting something outside its node.

**Comment by iseulde** [comment]
Fix will be released Monday: https://wordpress.slack.com/archives/C0UCMQP0F/p1507294443000422.

**Comment by iseulde** [comment]
Updated once more: https://github.com/tinymce/tinymce/compare/4.7.0...4.7.1

**Review by aduth — APPROVED** [review]
Updated `package-lock.json` to latest version. Confirmed previous issue I saw is no longer present. Appears to work well 👍

**Comment by iseulde** [comment]
> Updated package-lock.json

Sorry, again forgot 🙈
